pkg/controllers/dynakube/controller.go

```diff
@@ -318,11 +317,23 @@ func (controller *Controller) setupTokensAndClient(ctx context.Context, dynakube
 	return dynatraceClient, nil
 }
 
-func (controller *Controller) reconcileComponents(ctx context.Context, dynatraceClient dtclient.Client, istioReconciler istio.Reconciler, dynakube *dynatracev1beta1.DynaKube) error {
+func (controller *Controller) reconcileComponents(ctx context.Context, dynatraceClient dtclient.Client, istioClient *istio.Client, dynakube *dynatracev1beta1.DynaKube) error {
 	// it's important to setup app injection before AG so that it is already working when AG pods start, in case code modules shall get
 	// injected into AG for self-monitoring reasons
 	versionReconciler := controller.versionReconcilerBuilder(controller.apiReader, dynatraceClient, controller.fs, timeprovider.New().Freeze())
 	connectionInfoReconciler := controller.connectionInfoReconcilerBuilder(controller.client, controller.apiReader, controller.scheme, dynatraceClient)
+	injectionReconciler := controller.injectionReconcilerBuilder(controller.client, dynatraceClient, istioClient, controller.apiReader, controller.fs, dynakube)
+
+	var istioReconciler istio.Reconciler
+
+	if istioClient != nil {
+		istioReconciler = controller.istioReconcilerBuilder(istioClient)
+
+		err := istioReconciler.ReconcileAPIUrl(ctx, dynakube)
+		if err != nil {
+			return errors.WithMessage(err, "failed to reconcile istio objects for API url")
+		}
+	}
```

Comment:
this should be done earlier

the purpose of `ReconcileAPIUrl` is that the operator can talk to the Dynatrace API even if the operator's namespace is restricted with istio.

So we have to do this BEFORE ever using the dtclient.
- and even the creation of the dtclient requires the use of the dtclient, because we check the scopes at the same time [(kinda dumb, but thats how it works 😅)](https://github.com/Dynatrace/dynatrace-operator/blob/137c5b0b89ed633b11525a537a1bea723f9dddfa/pkg/controllers/dynakube/dynatraceclient/builder.go#L98)

this is why `setupIstio` is at the very beginning of the reconcile

Reply:
[Fixed here](https://github.com/Dynatrace/dynatrace-operator/pull/2736/commits/c0e8fd6e775fbe176dd1499d60be3f5f61b33e2f)